libraries/AP_Baro/AP_Baro_SPL06.cpp

```diff
@@ -0,0 +1,234 @@
+#define SPL06_MEAS_CFG_SENSOR_RDY              (1<<6)
+#define SPL06_MEAS_CFG_COEFFS_RDY              (1<<7)
+
+// INT_AND_FIFO_CFG_REG
+#define SPL06_PRESSURE_RESULT_BIT_SHIFT        (1<<2)  // necessary for pressure oversampling > 8
+#define SPL06_TEMPERATURE_RESULT_BIT_SHIFT     (1<<3)  // necessary for temperature oversampling > 8
+
+// Don't set oversampling higher than 8 or the measurement time will be higher than 20ms (timer period)
+#define SPL06_PRESSURE_OVERSAMPLING            8
+#define SPL06_TEMPERATURE_OVERSAMPLING         8
+
+#define SPL06_OVERSAMPLING_TO_REG_VALUE(n)     (ffs(n)-1)
+
+AP_Baro_SPL06::AP_Baro_SPL06(AP_Baro &baro, AP_HAL::OwnPtr<AP_HAL::Device> dev)
+    : AP_Baro_Backend(baro)
+    , _dev(std::move(dev))
+{
+}
+
+AP_Baro_Backend *AP_Baro_SPL06::probe(AP_Baro &baro,
+                                       AP_HAL::OwnPtr<AP_HAL::Device> dev)
+{
+    if (!dev) {
+        return nullptr;
+    }
+
+    if (dev->bus_type() == AP_HAL::Device::BUS_TYPE_SPI) {
+        dev->set_read_flag(0x80);
+    }
+
+    AP_Baro_SPL06 *sensor = new AP_Baro_SPL06(baro, std::move(dev));
+    if (!sensor || !sensor->_init()) {
+        delete sensor;
+        return nullptr;
+    }
+    return sensor;
+}
+
+bool AP_Baro_SPL06::_init()
+{
+    if (!_dev) {
+        return false;
+    }
+    WITH_SEMAPHORE(_dev->get_semaphore());
+
+    _has_sample = false;
+
+    _dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+
+    uint8_t whoami;
+    if (!_dev->read_registers(SPL06_REG_CHIP_ID, &whoami, 1)  ||
+        whoami != SPL06_CHIP_ID) {
+        // not a SPL06
+        return false;
+    }
+
+    // read the calibration data
+    uint8_t buf[SPL06_CALIB_COEFFS_LEN];
+    _dev->read_registers(SPL06_REG_CALIB_COEFFS_START, buf, sizeof(buf));
+
+    _c0 = (buf[0] & 0x80 ? 0xF000 : 0) | ((uint16_t)buf[0] << 4) | (((uint16_t)buf[1] & 0xF0) >> 4);
+    _c1 = ((buf[1] & 0x8 ? 0xF000 : 0) | ((uint16_t)buf[1] & 0x0F) << 8) | (uint16_t)buf[2];
+    _c00 = (buf[3] & 0x80 ? 0xFFF00000 : 0) | ((uint32_t)buf[3] << 12) | ((uint32_t)buf[4] << 4) | (((uint32_t)buf[5] & 0xF0) >> 4);
+    _c10 = (buf[5] & 0x8 ? 0xFFF00000 : 0) | (((uint32_t)buf[5] & 0x0F) << 16) | ((uint32_t)buf[6] << 8) | (uint32_t)buf[7];
+    _c01 = ((uint16_t)buf[8] << 8) | ((uint16_t)buf[9]);
+    _c11 = ((uint16_t)buf[10] << 8) | (uint16_t)buf[11];
+    _c20 = ((uint16_t)buf[12] << 8) | (uint16_t)buf[13];
+    _c21 = ((uint16_t)buf[14] << 8) | (uint16_t)buf[15];
+    _c30 = ((uint16_t)buf[16] << 8) | (uint16_t)buf[17];
+
+    // setup temperature and pressure measurements
+    _dev->setup_checked_registers(3, 20);
+
+    _dev->write_register(SPL06_REG_TEMPERATURE_CFG, SPL06_TEMP_USE_EXT_SENSOR | SPL06_OVERSAMPLING_TO_REG_VALUE(SPL06_TEMPERATURE_OVERSAMPLING), true);
+    _dev->write_register(SPL06_REG_PRESSURE_CFG, SPL06_OVERSAMPLING_TO_REG_VALUE(SPL06_PRESSURE_OVERSAMPLING), true);
+
+    uint8_t int_and_fifo_reg_value = 0;
+    if (SPL06_TEMPERATURE_OVERSAMPLING > 8) {
```

Comment:
isn't this dead code?

Reply:
It is dead code in this case but won't be if for any reason the oversampling value is raised